src/server/frontend_wayland/idle_inhibit_v1.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Grayson Guarino <grayson.guarino@canonical.com>
+ */
+
+#include "idle_inhibit_v1.h"
+
+#include "mir/log.h"
+#include "mir/executor.h"
+#include "mir/scene/surface.h"
+#include "mir/scene/null_surface_observer.h"
+
+#include "wl_surface.h"
+
+#include <boost/throw_exception.hpp>
+#include <memory>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace ms = mir::scene;
+
+namespace mir
+{
+namespace frontend
+{
+struct IdleInhibitV1Ctx
+{
+    class SurfaceObserver : ms::NullSurfaceObserver
+    {
+    public:
+        SurfaceObserver()
+        {
+        }
+
+        ~SurfaceObserver()
+        {
+        }
+
+        void attrib_changed(const ms::Surface *surf, MirWindowAttrib attrib, int value) override
+        {
+            (void)surf;
+            if (attrib == mir_window_attrib_focus && value == mir_window_focus_state_unfocused)
+                ; // TODO
+            else if (attrib == mir_window_attrib_focus)
+                ; // TODO
+        };
+    };
+
+    std::shared_ptr<Executor> const wayland_executor;
+    std::shared_ptr<ms::IdleHub> const idle_hub;
+    std::shared_ptr<ms::SurfaceObserver> surface_observer;
+};
+
+class IdleInhibitManagerV1Global
+    : public wayland::IdleInhibitManagerV1::Global
+{
+public:
+    IdleInhibitManagerV1Global(
+            wl_display* display,
+            std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+
+private:
+    void bind(wl_resource* new_resource) override;
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+};
+
+class IdleInhibitManagerV1
+    : public wayland::IdleInhibitManagerV1
+{
+public:
+    IdleInhibitManagerV1(
+            wl_resource* resource,
+            std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+
+
+private:
+    void create_inhibitor(struct wl_resource* id, struct wl_resource* surface) override;
+
+    std::shared_ptr<IdleInhibitV1Ctx> const& ctx;
```

Comment:
Yes, passing around reference to smart pointers often makes sense, but rarely do you want to store them. If you know for sure the object will outlive the reference than a reference directly to the pointed at object makes more sense, and if you want to hold the object alive or safely cope with it being destroyed a smart or weak pointer respectively (not a reference to those types) is required.

Reply:
> Yes, passing around reference to smart pointers often makes sense

MUCH less so since the incorporation of move semantics into the language and library.

The normal idiom should now be copy (or move) into a call or constructor and then move (or swap) as needed.

Reply:
This is still storing a dangling reference:
```c++
    auto ctx = std::make_shared<IdleInhibitV1Ctx>(IdleInhibitV1Ctx{wayland_executor, idle_hub});
    return std::make_shared<IdleInhibitManagerV1Global>(display, ctx);
```
`ctx` goes out of scope after the `return`.

Reply:
I want `IdleInhibitManagerV1Global` to own `ctx` so I followed your first suggestion and changed it to a non-reference.

Reply:
> The normal idiom should now be copy (or move) into a call or constructor and then move (or swap) as needed.

I'm not yet sold on that, we should discuss at some point.

Grayson It looks like you changed the constructor argument, but not the value held by the class. The important thing is that the class member is not a reference, or else as soon as the `shared_ptr` instance it refers to is dropped it will be invalid.